## Part I: Tensorflow

In [1]:
import os
import tensorflow as tf
import numpy as np
import math
import timeit
import matplotlib.pyplot as plt

%matplotlib inline

2023-10-11 15:22:59.142906: I tensorflow/core/util/port.cc:111] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2023-10-11 15:22:59.207158: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.
2023-10-11 15:22:59.511412: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:9342] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2023-10-11 15:22:59.511450: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2023-10-11 15:22:59.513095: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1518] Unable to register cuBLAS factory: Attempting to regi

In [2]:
def load_cifar10(num_training=49000, num_validation=1000, num_test=10000):
    """
    Fetch the CIFAR-10 dataset from the web and perform preprocessing to prepare
    it for the two-layer neural net classifier.
    """
    # Load the raw CIFAR-10 dataset and use appropriate data types and shapes
    cifar10 = tf.keras.datasets.cifar10.load_data()
    (X_train, y_train), (X_test, y_test) = cifar10
    X_train = np.asarray(X_train, dtype=np.float32)
    y_train = np.asarray(y_train, dtype=np.int32).flatten()
    X_test = np.asarray(X_test, dtype=np.float32)
    y_test = np.asarray(y_test, dtype=np.int32).flatten()

    # Subsample the data
    mask = range(num_training, num_training + num_validation)
    X_val = X_train[mask]
    y_val = y_train[mask]
    mask = range(num_training)
    X_train = X_train[mask]
    y_train = y_train[mask]
    mask = range(num_test)
    X_test = X_test[mask]
    y_test = y_test[mask]

    # Normalize the data: subtract the mean pixel and divide by std
    mean_pixel = X_train.mean(axis=(0, 1, 2), keepdims=True)
    std_pixel = X_train.std(axis=(0, 1, 2), keepdims=True)
    X_train = (X_train - mean_pixel) / std_pixel
    X_val = (X_val - mean_pixel) / std_pixel
    X_test = (X_test - mean_pixel) / std_pixel

    return X_train, y_train, X_val, y_val, X_test, y_test

# If there are errors with SSL downloading involving self-signed certificates,
# it may be that your Python version was recently installed on the current machine.
# See: https://github.com/tensorflow/tensorflow/issues/10779
# To fix, run the command: /Applications/Python\ 3.7/Install\ Certificates.command
#   ...replacing paths as necessary.

# Invoke the above function to get our data.
NHW = (0, 1, 2)
X_train, y_train, X_val, y_val, X_test, y_test = load_cifar10()
print('Train data shape: ', X_train.shape)
print('Train labels shape: ', y_train.shape, y_train.dtype)
print('Validation data shape: ', X_val.shape)
print('Validation labels shape: ', y_val.shape)
print('Test data shape: ', X_test.shape)
print('Test labels shape: ', y_test.shape)

Train data shape:  (49000, 32, 32, 3)
Train labels shape:  (49000,) int32
Validation data shape:  (1000, 32, 32, 3)
Validation labels shape:  (1000,)
Test data shape:  (10000, 32, 32, 3)
Test labels shape:  (10000,)


In [3]:
class Dataset(object):
    def __init__(self, X, y, batch_size, shuffle=False):
        """
        Construct a Dataset object to iterate over data X and labels y

        Inputs:
        - X: Numpy array of data, of any shape
        - y: Numpy array of labels, of any shape but with y.shape[0] == X.shape[0]
        - batch_size: Integer giving number of elements per minibatch
        - shuffle: (optional) Boolean, whether to shuffle the data on each epoch
        """
        assert X.shape[0] == y.shape[0], 'Got different numbers of data and labels'
        self.X, self.y = X, y
        self.batch_size, self.shuffle = batch_size, shuffle

    def __iter__(self):
        N, B = self.X.shape[0], self.batch_size
        idxs = np.arange(N)
        if self.shuffle:
            np.random.shuffle(idxs)
        return iter((self.X[i:i+B], self.y[i:i+B]) for i in range(0, N, B))


train_dset = Dataset(X_train, y_train, batch_size=64, shuffle=True)
val_dset = Dataset(X_val, y_val, batch_size=64, shuffle=False)
test_dset = Dataset(X_test, y_test, batch_size=64)

In [4]:
# We can iterate through a dataset like this:
for t, (x, y) in enumerate(train_dset):
    print(t, x.shape, y.shape)
    if t > 5: break

0 (64, 32, 32, 3) (64,)
1 (64, 32, 32, 3) (64,)
2 (64, 32, 32, 3) (64,)
3 (64, 32, 32, 3) (64,)
4 (64, 32, 32, 3) (64,)
5 (64, 32, 32, 3) (64,)
6 (64, 32, 32, 3) (64,)


In [5]:
# Set up some global variables
USE_GPU = False

if USE_GPU:
    device = '/device:GPU:0'
else:
    device = '/cpu:0'

# Constant to control how often we print when training models
print_every = 100

print('Using device: ', device)

Using device:  /cpu:0


In [6]:
## Helper Functions
def flatten(x):
    """
    Input:
    - TensorFlow Tensor of shape (N, D1, ..., DM)

    Output:
    - TensorFlow Tensor of shape (N, D1 * ... * DM)
    """
    N = tf.shape(x)[0]
    return tf.reshape(x, (N, -1))

def test_flatten():
    # Construct concrete values of the input data x using numpy
    x_np = np.arange(24).reshape((2, 3, 4))
    print('x_np:\n', x_np, '\n')
    # Compute a concrete output value.
    x_flat_np = flatten(x_np)
    print('x_flat_np:\n', x_flat_np, '\n')

test_flatten()

x_np:
 [[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]] 

x_flat_np:
 tf.Tensor(
[[ 0  1  2  3  4  5  6  7  8  9 10 11]
 [12 13 14 15 16 17 18 19 20 21 22 23]], shape=(2, 12), dtype=int64) 



### Define a TwoLayer Network

We will now implement our first neural network with TensorFlow: a fully-connected ReLU network with two hidden layers and no biases on the CIFAR10 dataset. For now we will use only low-level TensorFlow operators to define the network; later we will see how to use the higher-level abstractions provided by tf.keras to simplify the process.

We will define the forward pass of the network in the function TwoLayerFCNeuralNetwork; this will accept TensorFlow Tensors for the inputs and weights of the network, and return a TensorFlow Tensor for the scores.

After defining the network architecture in the TwoLayerFCNeuralNetwork function, we will test the implementation by checking the shape of the output.

It's important that you read and understand this implementation.

In [7]:
def TwoLayerFCNeuralNetwork(x, params):
    """
    A fully-connected neural network; the architecture is:
    fully-connected layer -> ReLU -> fully connected layer.
    Note that we only need to define the forward pass here; TensorFlow will take
    care of computing the gradients for us.

    The input to the network will be a minibatch of data, of shape
    (N, d1, ..., dM) where d1 * ... * dM = D. The hidden layer will have H units,
    and the output layer will produce scores for C classes.

    Inputs:
    - x: A TensorFlow Tensor of shape (N, d1, ..., dM) giving a minibatch of
      input data.
    - params: A list [w1, w2] of TensorFlow Tensors giving weights for the
      network, where w1 has shape (D, H) and w2 has shape (H, C).

    Returns:
    - scores: A TensorFlow Tensor of shape (N, C) giving classification scores
      for the input data x.
    """
    w1, w2 = params                   # Unpack the parameters
    x = flatten(x)                    # Flatten the input; now x has shape (N, D)
    h = tf.nn.relu(tf.matmul(x, w1))  # Hidden layer: h has shape (N, H)
    scores = tf.matmul(h, w2)         # Compute scores of shape (N, C)
    return scores

In [8]:
def TwoLayerFCNeuralNetwork_test():
    hidden_layer_size = 42

    # Scoping our TF operations under a tf.device context manager
    # lets us tell TensorFlow where we want these Tensors to be
    # multiplied and/or operated on, e.g. on a CPU or a GPU.
    with tf.device(device):
        x = tf.zeros((64, 32, 32, 3))
        w1 = tf.zeros((32 * 32 * 3, hidden_layer_size))
        w2 = tf.zeros((hidden_layer_size, 10))

        # Call our TwoLayerFCNeuralNetwork function for the forward pass of the network.
        scores = TwoLayerFCNeuralNetwork(x, [w1, w2])

    print(scores.shape)

TwoLayerFCNeuralNetwork_test()

(64, 10)


## Three-Layer ConvNet
Here you will complete the implementation of the function ThreeLayerConvNet which will perform the forward pass of a three-layer convolutional network. The network should have the following architecture:

A convolutional layer (with bias) with channel_1 filters, each with shape KW1 x KH1, and zero-padding of two
ReLU nonlinearity
A convolutional layer (with bias) with channel_2 filters, each with shape KW2 x KH2, and zero-padding of one
ReLU nonlinearity
Fully-connected layer with bias, producing scores for C classes.

HINT: For convolutions: https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/nn/conv2d; be careful with padding!

HINT: For biases: https://www.tensorflow.org/performance/xla/broadcasting

In [9]:
def ThreeLayerConvNet(x, params):
    """
    A three-layer convolutional network with the architecture described above.

    Inputs:
    - x: A TensorFlow Tensor of shape (N, H, W, 3) giving a minibatch of images
    - params: A list of TensorFlow Tensors giving the weights and biases for the
      network; should contain the following:
      - conv_w1: TensorFlow Tensor of shape (KH1, KW1, 3, channel_1) giving
        weights for the first convolutional layer.
      - conv_b1: TensorFlow Tensor of shape (channel_1,) giving biases for the
        first convolutional layer.
      - conv_w2: TensorFlow Tensor of shape (KH2, KW2, channel_1, channel_2)
        giving weights for the second convolutional layer
      - conv_b2: TensorFlow Tensor of shape (channel_2,) giving biases for the
        second convolutional layer.
      - fc_w: TensorFlow Tensor giving weights for the fully-connected layer.
        Can you figure out what the shape should be?
      - fc_b: TensorFlow Tensor giving biases for the fully-connected layer.
        Can you figure out what the shape should be?
    """
    conv_w1, conv_b1, conv_w2, conv_b2, fc_w, fc_b = params
    scores = None
    ############################################################################
    # TODO: Implement the forward pass for the three-layer ConvNet.            #
    ############################################################################
    # *****START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
   
    # First convolutional layer
    conv1 = tf.nn.conv2d(x, conv_w1, strides=[1, 1, 1, 1], padding='SAME')
    conv1 = tf.nn.bias_add(conv1, conv_b1)
    conv1 = tf.nn.relu(conv1)

    # Second convolutional layer
    conv2 = tf.nn.conv2d(conv1, conv_w2, strides=[1, 1, 1, 1], padding='SAME')
    conv2 = tf.nn.bias_add(conv2, conv_b2)
    conv2 = tf.nn.relu(conv2)

    # Fully connected layer
    N, H, W, C = conv2.shape
    flattened = tf.reshape(conv2, (N, -1))  # Flatten the feature maps
    fc = tf.matmul(flattened, fc_w) + fc_b
    scores = fc

    # *****END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    ############################################################################
    #                              END OF YOUR CODE                            #
    ############################################################################
    return scores

After defing the forward pass of the three-layer ConvNet above, run the following cell to test your implementation. Like the two-layer network, we run the graph on a batch of zeros just to make sure the function doesn't crash, and produces outputs of the correct shape.

When you run this function, scores_np should have shape (64, 10).

In [10]:
def ThreeLayerConvNetTest():

    with tf.device(device):
        x = tf.zeros((64, 32, 32, 3))
        conv_w1 = tf.zeros((5, 5, 3, 6))
        conv_b1 = tf.zeros((6,))
        conv_w2 = tf.zeros((3, 3, 6, 9))
        conv_b2 = tf.zeros((9,))
        fc_w = tf.zeros((32 * 32 * 9, 10))
        fc_b = tf.zeros((10,))
        params = [conv_w1, conv_b1, conv_w2, conv_b2, fc_w, fc_b]
        scores = ThreeLayerConvNet(x, params)

    # Inputs to convolutional layers are 4-dimensional arrays with shape
    # [batch_size, height, width, channels]
    print('scores_np has shape: ', scores.shape)

ThreeLayerConvNetTest()

scores_np has shape:  (64, 10)


### Training Step
We now define the training_step function performs a single training step. This will take three basic steps:

Compute the loss
Compute the gradient of the loss with respect to all network weights
Make a weight update step using (stochastic) gradient descent.
We need to use a few new TensorFlow functions to do all of this:

For computing the cross-entropy loss we'll use tf.nn.
sparse_softmax_cross_entropy_with_logits: https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/nn/sparse_softmax_cross_entropy_with_logits

For averaging the loss across a minibatch of data we'll use tf.reduce_mean:

https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/reduce_mean

For computing gradients of the loss with respect to the weights we'll use tf.GradientTape (useful for Eager execution): https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/GradientTape

We'll mutate the weight values stored in a TensorFlow Tensor using tf.compat.v1.assign_sub ("sub" is for subtraction): https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/compat/v1/assign_sub

In [11]:
def trainingStep(model_fn, x, y, params, learning_rate):
    with tf.GradientTape() as tape:
        scores = model_fn(x, params) # Forward pass of the model
        loss = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=y, logits=scores)
        total_loss = tf.reduce_mean(loss)
        grad_params = tape.gradient(total_loss, params)

        # Make a vanilla gradient descent step on all of the model parameters
        # Manually update the weights using assign_sub()
        for w, grad_w in zip(params, grad_params):
            w.assign_sub(learning_rate * grad_w)

        return total_loss

In [12]:
def trainingStep2(model_fn, init_fn, learning_rate):
    """
    Train a model on CIFAR-10.

    Inputs:
    - model_fn: A Python function that performs the forward pass of the model
      using TensorFlow; it should have the following signature:
      scores = model_fn(x, params) where x is a TensorFlow Tensor giving a
      minibatch of image data, params is a list of TensorFlow Tensors holding
      the model weights, and scores is a TensorFlow Tensor of shape (N, C)
      giving scores for all elements of x.
    - init_fn: A Python function that initializes the parameters of the model.
      It should have the signature params = init_fn() where params is a list
      of TensorFlow Tensors holding the (randomly initialized) weights of the
      model.
    - learning_rate: Python float giving the learning rate to use for SGD.
    """
    params = init_fn()  # Initialize the model parameters

    for t, (x_np, y_np) in enumerate(train_dset):
        # Run the graph on a batch of training data.
        loss = trainingStep(model_fn, x_np, y_np, params, learning_rate)

        # Periodically print the loss and check accuracy on the val set.
        if t % print_every == 0:
            print('Iteration %d, loss = %.4f' % (t, loss))
            check_accuracy(val_dset, x_np, model_fn, params)

In [13]:
def check_accuracy(dset, x, model_fn, params):
    """
    Check accuracy on a classification model, e.g. for validation.

    Inputs:
    - dset: A Dataset object against which to check accuracy
    - x: A TensorFlow placeholder Tensor where input images should be fed
    - model_fn: the Model we will be calling to make predictions on x
    - params: parameters for the model_fn to work with

    Returns: Nothing, but prints the accuracy of the model
    """
    num_correct, num_samples = 0, 0
    for x_batch, y_batch in dset:
        scores_np = model_fn(x_batch, params).numpy()
        y_pred = scores_np.argmax(axis=1)
        num_samples += x_batch.shape[0]
        num_correct += (y_pred == y_batch).sum()
    acc = float(num_correct) / num_samples
    print('Got %d / %d correct (%.2f%%)' % (num_correct, num_samples, 100 * acc))

### Initialization
We'll use the following utility method to initialize the weight matrices for our models using Kaiming's normalization method.

[1] He et al, Delving Deep into Rectifiers: Surpassing Human-Level Performance on ImageNet Classification , ICCV 2015, https://arxiv.org/abs/1502.01852

In [14]:
def create_matrix_with_kaiming_normal(shape):
    if len(shape) == 2:
        fan_in, fan_out = shape[0], shape[1]
    elif len(shape) == 4:
        fan_in, fan_out = np.prod(shape[:3]), shape[3]
    return tf.keras.backend.random_normal(shape) * np.sqrt(2.0 / fan_in)

### Train a Two-Layer Network
We are finally ready to use all of the pieces defined above to train a two-layer fully-connected network on CIFAR-10.

We just need to define a function to initialize the weights of the model, and call train_part2.

Defining the weights of the network introduces another important piece of TensorFlow API: tf.Variable. A TensorFlow Variable is a Tensor whose value is stored in the graph and persists across runs of the computational graph; however unlike constants defined with tf.zeros or tf.random_normal, the values of a Variable can be mutated as the graph runs; these mutations will persist across graph runs. Learnable parameters of the network are usually stored in Variables.

You don't need to tune any hyperparameters, but you should achieve validation accuracies above 40% after one epoch of training.

In [15]:
def TwoLayerFCNeuralNetwork_init():
    """
    Initialize the weights of a two-layer network, for use with the
    TwoLayerFCNeuralNetwork function defined above.
    You can use the `create_matrix_with_kaiming_normal` helper!

    Inputs: None

    Returns: A list of:
    - w1: TensorFlow tf.Variable giving the weights for the first layer
    - w2: TensorFlow tf.Variable giving the weights for the second layer
    """
    hidden_layer_size = 4000
    w1 = tf.Variable(create_matrix_with_kaiming_normal((3 * 32 * 32, 4000)))
    w2 = tf.Variable(create_matrix_with_kaiming_normal((4000, 10)))
    return [w1, w2]
learning_rate = 1e-2
trainingStep2(TwoLayerFCNeuralNetwork, TwoLayerFCNeuralNetwork_init, learning_rate)

Iteration 0, loss = 3.1796
Got 124 / 1000 correct (12.40%)
Iteration 100, loss = 1.8276
Got 366 / 1000 correct (36.60%)
Iteration 200, loss = 1.4889
Got 382 / 1000 correct (38.20%)
Iteration 300, loss = 1.7739
Got 365 / 1000 correct (36.50%)
Iteration 400, loss = 1.8209
Got 417 / 1000 correct (41.70%)
Iteration 500, loss = 1.7822
Got 434 / 1000 correct (43.40%)
Iteration 600, loss = 1.7922
Got 418 / 1000 correct (41.80%)
Iteration 700, loss = 1.9471
Got 418 / 1000 correct (41.80%)


### Train a three-layer ConvNet
We will now use TensorFlow to train a three-layer ConvNet on CIFAR-10.

You need to implement the ThreeLayerConvNet_init function. Recall that the architecture of the network is:

Convolutional layer (with bias) with 32 5x5 filters, with zero-padding 2
ReLU
Convolutional layer (with bias) with 16 3x3 filters, with zero-padding 1
ReLU
Fully-connected layer (with bias) to compute scores for 10 classes
You don't need to do any hyperparameter tuning, but you should see validation accuracies above 43% after one epoch of training.


In [16]:
def ThreeLayerConvNet_init():
    """
    Initialize the weights of a Three-Layer ConvNet, for use with the
    ThreeLayerConvNet function defined above.
    You can use the `create_matrix_with_kaiming_normal` helper!

    Inputs: None

    Returns a list containing:
    - conv_w1: TensorFlow tf.Variable giving weights for the first conv layer
    - conv_b1: TensorFlow tf.Variable giving biases for the first conv layer
    - conv_w2: TensorFlow tf.Variable giving weights for the second conv layer
    - conv_b2: TensorFlow tf.Variable giving biases for the second conv layer
    - fc_w: TensorFlow tf.Variable giving weights for the fully-connected layer
    - fc_b: TensorFlow tf.Variable giving biases for the fully-connected layer
    """
    params = None
    ############################################################################
    # TODO: Initialize the parameters of the three-layer network.              #
    ############################################################################
    # *****START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    
    num_classes = 10
    H_out, W_out = 32, 32
    channel_1, channel_2 = 32, 16
    filter_1, filter_2 = (5, 5), (3, 3)
    
    # Initialize weights for the first convolutional layer
    conv_w1 = tf.Variable(create_matrix_with_kaiming_normal((*filter_1, 3, channel_1)))
    conv_b1 = tf.Variable(tf.zeros(channel_1))
    
    # Initialize weights for the second convolutional layer
    conv_w2 = tf.Variable(create_matrix_with_kaiming_normal((*filter_2, channel_1, channel_2)))
    conv_b2 = tf.Variable(tf.zeros(channel_2))
    
    # Initialize weights for the fully connected layer
    fc_w = tf.Variable(create_matrix_with_kaiming_normal((H_out * W_out * channel_2, num_classes)))
    fc_b = tf.Variable(tf.zeros(num_classes))
    
    params = [conv_w1, conv_b1, conv_w2, conv_b2, fc_w, fc_b]

    # *****END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    ############################################################################
    #                             END OF YOUR CODE                             #
    ############################################################################
    return params

learning_rate = 3e-3
trainingStep2(ThreeLayerConvNet, ThreeLayerConvNet_init, learning_rate)

Iteration 0, loss = 2.8774
Got 88 / 1000 correct (8.80%)
Iteration 100, loss = 1.8314
Got 358 / 1000 correct (35.80%)
Iteration 200, loss = 1.5761
Got 387 / 1000 correct (38.70%)
Iteration 300, loss = 1.6778
Got 382 / 1000 correct (38.20%)
Iteration 400, loss = 1.5478
Got 420 / 1000 correct (42.00%)
Iteration 500, loss = 1.7178
Got 446 / 1000 correct (44.60%)
Iteration 600, loss = 1.6588
Got 460 / 1000 correct (46.00%)
Iteration 700, loss = 1.6797
Got 462 / 1000 correct (46.20%)


## Part III: Keras Model Subclassing API
Implementing a neural network using the low-level TensorFlow API is a good way to understand how TensorFlow works, but it's a little inconvenient - we had to manually keep track of all Tensors holding learnable parameters. This was fine for a small network, but could quickly become unweildy for a large complex model.

Fortunately TensorFlow 2.0 provides higher-level APIs such as tf.keras which make it easy to build models out of modular, object-oriented layers. Further, TensorFlow 2.0 uses eager execution that evaluates operations immediately, without explicitly constructing any computational graphs. This makes it easy to write and debug models, and reduces the boilerplate code.

In this part of the notebook we will define neural network models using the tf.keras.Model API. To implement your own model, you need to do the following:

Define a new class which subclasses tf.keras.Model. Give your class an intuitive name that describes it, like twoLayerFC or threeLayerConvNet.
In the initializer __init__() for your new class, define all the layers you need as class attributes. The tf.keras.layers package provides many common neural-network layers, like tf.keras.layers.Dense for fully-connected layers and tf.keras.layers.Conv2D for convolutional layers. Under the hood, these layers will construct Variable Tensors for any learnable parameters. Warning: Don't forget to call super(YourModelName, self).__init__() as the first line in your initializer!
Implement the call() method for your class; this implements the forward pass of your model, and defines the connectivity of your network. Layers defined in __init__() implement __call__() so they can be used as function objects that transform input Tensors into output Tensors. Don't define any new layers in call(); any layers you want to use in the forward pass should be defined in __init__().
After you define your tf.keras.Model subclass, you can instantiate it and use it like the model functions from Part II.

Keras Model Subclassing API: Two-Layer Network
Here is a concrete example of using the tf.keras.Model API to define a two-layer network. There are a few new bits of API to be aware of here:

We use an Initializer object to set up the initial values of the learnable parameters of the layers; in particular tf.initializers.VarianceScaling gives behavior similar to the Kaiming initialization method we used in Part II. You can read more about it here: https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/initializers/VarianceScaling

We construct tf.keras.layers.Dense objects to represent the two fully-connected layers of the model. In addition to multiplying their input by a weight matrix and adding a bias vector, these layer can also apply a nonlinearity for you. For the first layer we specify a ReLU activation function by passing activation='relu' to the constructor; the second layer uses softmax activation function. Finally, we use tf.keras.layers.Flatten to flatten the output from the previous fully-connected layer.

In [17]:
class twoLayerFC(tf.keras.Model):
    def __init__(self, hidden_size, num_classes):
        super(twoLayerFC, self).__init__()
        initializer = tf.initializers.VarianceScaling(scale=2.0)
        self.fc1 = tf.keras.layers.Dense(hidden_size, activation='relu',
                                   kernel_initializer=initializer)
        self.fc2 = tf.keras.layers.Dense(num_classes, activation='softmax',
                                   kernel_initializer=initializer)
        self.flatten = tf.keras.layers.Flatten()

    def call(self, x, training=False):
        x = self.flatten(x)
        x = self.fc1(x)
        x = self.fc2(x)
        return x


def twoLayerFC_test():
    """ A small unit test to exercise the TwoLayerFC model above. """
    input_size, hidden_size, num_classes = 50, 42, 10
    x = tf.zeros((64, input_size))
    model = twoLayerFC(hidden_size, num_classes)
    with tf.device(device):
        scores = model(x)
        print(scores.shape)

twoLayerFC_test()

(64, 10)


/home/dscerbo/miniconda3/lib/python3.11/site-packages/keras/src/initializers/initializers.py:120: UserWarning: The initializer VarianceScaling is unseeded and being called multiple times, which will return identical values each time (even if the initializer is unseeded). Please update your code to provide a seed to the initializer, or avoid using the same initializer instance more than once.
  warnings.warn(


### Keras Model Subclassing API: Three-Layer ConvNet
Now it's your turn to implement a three-layer ConvNet using the tf.keras.Model API. Your model should have the same architecture used in Part II:

Convolutional layer with 5 x 5 kernels, with zero-padding of 2
ReLU nonlinearity
Convolutional layer with 3 x 3 kernels, with zero-padding of 1
ReLU nonlinearity
Fully-connected layer to give class scores
Softmax nonlinearity
You should initialize the weights of your network using the same initialization method as was used in the two-layer network above.

Hint: Refer to the documentation for tf.keras.layers.Conv2D and tf.keras.layers.Dense:

https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/layers/Conv2D

https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/layers/Dense

In [18]:
class threeLayerConvNet(tf.keras.Model):
    def __init__(self, channel_1, channel_2, num_classes):
        super(threeLayerConvNet, self).__init__()
        ########################################################################
        # TODO: Implement the __init__ method for a three-layer ConvNet. You   #
        # should instantiate layer objects to be used in the forward pass.     #
        ########################################################################
        # *****START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
        
        initializer = tf.initializers.VarianceScaling(scale=2.0)
        self.conv1 = tf.keras.layers.Conv2D(channel_1, (5, 5), 
                                            padding='same', activation='relu', 
                                            kernel_initializer=initializer)
        
        self.conv2 = tf.keras.layers.Conv2D(channel_2, (3, 3), 
                                            padding='same', activation='relu', 
                                            kernel_initializer=initializer)
        
        self.flatten = tf.keras.layers.Flatten()
        self.fc = tf.keras.layers.Dense(num_classes, activation='softmax',
                                        kernel_initializer=initializer)

        # *****END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
        ########################################################################
        #                           END OF YOUR CODE                           #
        ########################################################################

    def call(self, x, training=False):
        scores = None
        ########################################################################
        # TODO: Implement the forward pass for a three-layer ConvNet. You      #
        # should use the layer objects defined in the __init__ method.         #
        ########################################################################
        # *****START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****

        x = self.conv1(x)
        x = self.conv2(x)
        x = self.flatten(x)
        scores = self.fc(x)
        
        # *****END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
        ########################################################################
        #                           END OF YOUR CODE                           #
        ########################################################################
        return scores

Once you complete the implementation of the threeLayerConvNet above you can run the following to ensure that your implementation does not crash and produces outputs of the expected shape.

In [19]:
def threeLayerConvNet_test():
    channel_1, channel_2, num_classes = 12, 8, 10
    model = threeLayerConvNet(channel_1, channel_2, num_classes)
    with tf.device(device):
        x = tf.zeros((64, 3, 32, 32))
        scores = model(x)
        print(scores.shape)

threeLayerConvNet_test()

(64, 10)


### Keras Model Subclassing API: Eager Training
While keras models have a builtin training loop (using the model.fit), sometimes you need more customization. Here's an example, of a training loop implemented with eager execution.

In particular, notice tf.GradientTape. Automatic differentiation is used in the backend for implementing backpropagation in frameworks like TensorFlow. During eager execution, tf.GradientTape is used to trace operations for computing gradients later. A particular tf.GradientTape can only compute one gradient; subsequent calls to tape will throw a runtime error.

TensorFlow 2.0 ships with easy-to-use built-in metrics under tf.keras.metrics module. Each metric is an object, and we can use update_state() to add observations and reset_state() to clear all observations. We can get the current result of a metric by calling result() on the metric object.

In [20]:
def training_part3(model_init_fn, optimizer_init_fn, num_epochs=1, is_training=False):
    """
    Simple training loop for use with models defined using tf.keras. It trains
    a model for one epoch on the CIFAR-10 training set and periodically checks
    accuracy on the CIFAR-10 validation set.

    Inputs:
    - model_init_fn: A function that takes no parameters; when called it
      constructs the model we want to train: model = model_init_fn()
    - optimizer_init_fn: A function which takes no parameters; when called it
      constructs the Optimizer object we will use to optimize the model:
      optimizer = optimizer_init_fn()
    - num_epochs: The number of epochs to train for

    Returns: Nothing, but prints progress during trainingn
    """
    with tf.device(device):

        # Compute the loss like we did in Part II
        loss_fn = tf.keras.losses.SparseCategoricalCrossentropy()

        model = model_init_fn()
        optimizer = optimizer_init_fn()

        train_loss = tf.keras.metrics.Mean(name='train_loss')
        train_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(name='train_accuracy')

        val_loss = tf.keras.metrics.Mean(name='val_loss')
        val_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(name='val_accuracy')

        t = 0
        for epoch in range(num_epochs):

            # Reset the metrics - https://www.tensorflow.org/alpha/guide/migration_guide#new-style_metrics
            train_loss.reset_states()
            train_accuracy.reset_states()

            for x_np, y_np in train_dset:
                with tf.GradientTape() as tape:

                    # Use the model function to build the forward pass.
                    scores = model(x_np, training=is_training)
                    loss = loss_fn(y_np, scores)

                    gradients = tape.gradient(loss, model.trainable_variables)
                    optimizer.apply_gradients(zip(gradients, model.trainable_variables))

                    # Update the metrics
                    train_loss.update_state(loss)
                    train_accuracy.update_state(y_np, scores)

                    if t % print_every == 0:
                        val_loss.reset_states()
                        val_accuracy.reset_states()
                        for test_x, test_y in val_dset:
                            # During validation at end of epoch, training set to False
                            prediction = model(test_x, training=False)
                            t_loss = loss_fn(test_y, prediction)

                            val_loss.update_state(t_loss)
                            val_accuracy.update_state(test_y, prediction)

                        template = 'Iteration {}, Epoch {}, Loss: {:.4f}, Accuracy: {:.4f}, Val Loss: {:.4f}, Val Accuracy: {:.4f}'
                        print (template.format(t, epoch+1,
                                             train_loss.result(),
                                             train_accuracy.result()*100,
                                             val_loss.result(),
                                             val_accuracy.result()*100))
                    t += 1

### Keras Model Subclassing API: Train a Two-Layer Network
We can now use the tools defined above to train a two-layer network on CIFAR-10. We define the model_init_fn and optimizer_init_fn that construct the model and optimizer respectively when called. Here we want to train the model using stochastic gradient descent with no momentum, so we construct a tf.keras.optimizers.SGD function; you can read about it here.

You don't need to tune any hyperparameters here, but you should achieve validation accuracies above 40% after one epoch of training.

In [21]:
hidden_size, num_classes = 4000, 10
learning_rate = 1e-2

def model_init_fn():
    return twoLayerFC(hidden_size, num_classes)

def optimizer_init_fn():
    return tf.keras.optimizers.SGD(learning_rate=learning_rate)

training_part3(model_init_fn, optimizer_init_fn)

Iteration 0, Epoch 1, Loss: 3.0493, Accuracy: 9.3750, Val Loss: 2.8898, Val Accuracy: 13.6000
Iteration 100, Epoch 1, Loss: 2.2398, Accuracy: 28.3571, Val Loss: 1.9229, Val Accuracy: 36.4000
Iteration 200, Epoch 1, Loss: 2.0746, Accuracy: 32.1828, Val Loss: 1.8892, Val Accuracy: 38.3000
Iteration 300, Epoch 1, Loss: 2.0002, Accuracy: 34.0739, Val Loss: 1.8684, Val Accuracy: 37.9000
Iteration 400, Epoch 1, Loss: 1.9271, Accuracy: 36.0115, Val Loss: 1.7549, Val Accuracy: 40.9000
Iteration 500, Epoch 1, Loss: 1.8841, Accuracy: 37.0259, Val Loss: 1.6879, Val Accuracy: 42.0000
Iteration 600, Epoch 1, Loss: 1.8542, Accuracy: 37.9680, Val Loss: 1.7030, Val Accuracy: 41.3000
Iteration 700, Epoch 1, Loss: 1.8281, Accuracy: 38.6167, Val Loss: 1.6509, Val Accuracy: 44.1000


### Keras Model Subclassing API: Train a Three-Layer ConvNet
Here you should use the tools we've defined above to train a three-layer ConvNet on CIFAR-10. Your ConvNet should use 32 filters in the first convolutional layer and 16 filters in the second layer.

To train the model you should use gradient descent with Nesterov momentum 0.9.

HINT: https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/optimizers/SGD

You don't need to perform any hyperparameter tuning, but you should achieve validation accuracies above 50% after training for one epoch.

In [22]:
learning_rate = 3e-3
channel_1, channel_2, num_classes = 32, 16, 10

def model_init_fn():
    model = None
    ############################################################################
    # TODO: Complete the implementation of model_fn.                           #
    ############################################################################
    # *****START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****

    model = threeLayerConvNet(channel_1, channel_2, num_classes)

    # *****END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    ############################################################################
    #                           END OF YOUR CODE                               #
    ############################################################################
    return model

def optimizer_init_fn():
    optimizer = None
    ############################################################################
    # TODO: Complete the implementation of model_fn.                           #
    ############################################################################
    # *****START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****

    optimizer = tf.keras.optimizers.SGD(learning_rate=learning_rate, 
                                        momentum=0.9, nesterov=True)
    
    # *****END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    ############################################################################
    #                           END OF YOUR CODE                               #
    ############################################################################
    return optimizer

training_part3(model_init_fn, optimizer_init_fn)

Iteration 0, Epoch 1, Loss: 3.0440, Accuracy: 4.6875, Val Loss: 6.9287, Val Accuracy: 8.1000
Iteration 100, Epoch 1, Loss: 2.1411, Accuracy: 27.4134, Val Loss: 1.8235, Val Accuracy: 37.6000
Iteration 200, Epoch 1, Loss: 1.9400, Accuracy: 33.1468, Val Loss: 1.6038, Val Accuracy: 46.0000
Iteration 300, Epoch 1, Loss: 1.8242, Accuracy: 36.6123, Val Loss: 1.5408, Val Accuracy: 45.8000
Iteration 400, Epoch 1, Loss: 1.7339, Accuracy: 39.5496, Val Loss: 1.4484, Val Accuracy: 48.3000
Iteration 500, Epoch 1, Loss: 1.6706, Accuracy: 41.4577, Val Loss: 1.3807, Val Accuracy: 50.2000
Iteration 600, Epoch 1, Loss: 1.6259, Accuracy: 42.9753, Val Loss: 1.3452, Val Accuracy: 51.9000
Iteration 700, Epoch 1, Loss: 1.5872, Accuracy: 44.3429, Val Loss: 1.3262, Val Accuracy: 53.5000


## Part IV: Keras Sequential API
In Part III we introduced the tf.keras.Model API, which allows you to define models with any number of learnable layers and with arbitrary connectivity between layers.

However for many models you don't need such flexibility - a lot of models can be expressed as a sequential stack of layers, with the output of each layer fed to the next layer as input. If your model fits this pattern, then there is an even easier way to define your model: using tf.keras.Sequential. You don't need to write any custom classes; you simply call the tf.keras.Sequential constructor with a list containing a sequence of layer objects.

One complication with tf.keras.Sequential is that you must define the shape of the input to the model by passing a value to the input_shape of the first layer in your model.

Keras Sequential API: Two-Layer Network
In this subsection, we will rewrite the two-layer fully-connected network using tf.keras.Sequential, and train it using the training loop defined above.

You don't need to perform any hyperparameter tuning here, but you should see validation accuracies above 40% after training for one epoch.

In [23]:
learning_rate = 1e-2

def model_init_fn():
    input_shape = (32, 32, 3)
    hidden_layer_size, num_classes = 4000, 10
    initializer = tf.initializers.VarianceScaling(scale=2.0)
    layers = [
        tf.keras.layers.Flatten(input_shape=input_shape),
        tf.keras.layers.Dense(hidden_layer_size, activation='relu',
                              kernel_initializer=initializer),
        tf.keras.layers.Dense(num_classes, activation='softmax',
                              kernel_initializer=initializer),
    ]
    model = tf.keras.Sequential(layers)
    return model

def optimizer_init_fn():
    return tf.keras.optimizers.SGD(learning_rate=learning_rate)

training_part3(model_init_fn, optimizer_init_fn)

Iteration 0, Epoch 1, Loss: 3.0360, Accuracy: 6.2500, Val Loss: 3.1819, Val Accuracy: 12.3000
Iteration 100, Epoch 1, Loss: 2.2565, Accuracy: 28.5736, Val Loss: 1.9352, Val Accuracy: 38.0000
Iteration 200, Epoch 1, Loss: 2.0885, Accuracy: 32.0040, Val Loss: 1.8861, Val Accuracy: 39.5000
Iteration 300, Epoch 1, Loss: 2.0091, Accuracy: 33.6119, Val Loss: 1.9263, Val Accuracy: 37.2000
Iteration 400, Epoch 1, Loss: 1.9391, Accuracy: 35.7232, Val Loss: 1.7404, Val Accuracy: 43.0000
Iteration 500, Epoch 1, Loss: 1.8931, Accuracy: 36.8669, Val Loss: 1.6780, Val Accuracy: 43.4000
Iteration 600, Epoch 1, Loss: 1.8615, Accuracy: 37.7288, Val Loss: 1.6913, Val Accuracy: 42.2000
Iteration 700, Epoch 1, Loss: 1.8349, Accuracy: 38.4027, Val Loss: 1.6511, Val Accuracy: 43.7000


### Abstracting Away the Training Loop
In the previous examples, we used a customised training loop to train models (e.g. train_part34). Writing your own training loop is only required if you need more flexibility and control during training your model. Alternately, you can also use built-in APIs like tf.keras.Model.fit() and tf.keras.Model.evaluate to train and evaluate a model. Also remember to configure your model for training by calling `tf.keras.Model.compile.

You don't need to perform any hyperparameter tuning here, but you should see validation and test accuracies above 42% after training for one epoch.

In [24]:
model = model_init_fn()
model.compile(optimizer=tf.keras.optimizers.SGD(learning_rate=learning_rate),
              loss='sparse_categorical_crossentropy',
              metrics=[tf.keras.metrics.sparse_categorical_accuracy])
model.fit(X_train, y_train, batch_size=64, epochs=1, validation_data=(X_val, y_val))
model.evaluate(X_test, y_test)

313/313 [==============================] - 4s 13ms/step - loss: 1.6017 - sparse_categorical_accuracy: 0.4544


[1.6017005443572998, 0.4544000029563904]

### Keras Sequential API: Three-Layer ConvNet
Here you should use tf.keras.Sequential to reimplement the same three-layer ConvNet architecture used in Part II and Part III. As a reminder, your model should have the following architecture:

Convolutional layer with 32 5x5 kernels, using zero padding of 2
ReLU nonlinearity
Convolutional layer with 16 3x3 kernels, using zero padding of 1
ReLU nonlinearity
Fully-connected layer giving class scores
Softmax nonlinearity
You should initialize the weights of the model using a tf.initializers.VarianceScaling as above.

You should train the model using Nesterov momentum 0.9.

You don't need to perform any hyperparameter search, but you should achieve accuracy above 45% after training for one epoch.

In [25]:
def model_init_fn():
    model = None
    ############################################################################
    # TODO: Construct a three-layer ConvNet using tf.keras.Sequential.         #
    ############################################################################
    # *****START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    
    input_shape = (32, 32, 3)
    channel_1, channel_2, num_classes = 32, 16, 10
    initializer = tf.initializers.VarianceScaling(scale=2.0)
    layers = [
        tf.keras.layers.Conv2D(channel_1, (5, 5), 
                               padding='same', activation='relu', 
                               kernel_initializer=initializer,
                               input_shape=input_shape),
        tf.keras.layers.Conv2D(channel_2, (3, 3), 
                               padding='same', activation='relu', 
                               kernel_initializer=initializer),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(num_classes, activation='softmax',
                              kernel_initializer=initializer),
    ]
    model = tf.keras.Sequential(layers) 

    # *****END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    ############################################################################
    #                            END OF YOUR CODE                              #
    ############################################################################
    return model

learning_rate = 5e-4
def optimizer_init_fn():
    optimizer = None
    ############################################################################
    # TODO: Complete the implementation of model_fn.                           #
    ############################################################################
    # *****START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****

    optimizer = tf.keras.optimizers.SGD(learning_rate=learning_rate, 
                                        momentum=0.9, nesterov=True)

    # *****END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    ############################################################################
    #                           END OF YOUR CODE                               #
    ############################################################################
    return optimizer

training_part3(model_init_fn, optimizer_init_fn)

Iteration 0, Epoch 1, Loss: 2.8113, Accuracy: 7.8125, Val Loss: 2.6741, Val Accuracy: 7.7000
Iteration 100, Epoch 1, Loss: 1.9780, Accuracy: 29.5637, Val Loss: 1.7857, Val Accuracy: 37.9000
Iteration 200, Epoch 1, Loss: 1.8537, Accuracy: 34.6238, Val Loss: 1.6662, Val Accuracy: 43.3000
Iteration 300, Epoch 1, Loss: 1.7860, Accuracy: 37.1522, Val Loss: 1.6252, Val Accuracy: 44.2000
Iteration 400, Epoch 1, Loss: 1.7248, Accuracy: 39.1989, Val Loss: 1.5616, Val Accuracy: 46.1000
Iteration 500, Epoch 1, Loss: 1.6828, Accuracy: 40.7560, Val Loss: 1.5085, Val Accuracy: 47.3000
Iteration 600, Epoch 1, Loss: 1.6548, Accuracy: 41.8209, Val Loss: 1.4663, Val Accuracy: 50.1000
Iteration 700, Epoch 1, Loss: 1.6273, Accuracy: 42.8294, Val Loss: 1.4332, Val Accuracy: 52.4000


We will also train this model with the built-in training loop APIs provided by TensorFlow.

In [26]:
model = model_init_fn()
model.compile(optimizer='sgd',
              loss='sparse_categorical_crossentropy',
              metrics=[tf.keras.metrics.sparse_categorical_accuracy])
model.fit(X_train, y_train, batch_size=64, epochs=1, validation_data=(X_val, y_val))
model.evaluate(X_test, y_test)

313/313 [==============================] - 3s 10ms/step - loss: 1.3714 - sparse_categorical_accuracy: 0.5175


[1.3714126348495483, 0.5174999833106995]

### Part V: BONUS POINT - CIFAR-10 open-ended challenge
In this section you can experiment with whatever ConvNet architecture you'd like on CIFAR-10.

You should experiment with architectures, hyperparameters, loss functions, regularization, or anything else you can think of to train a model that achieves at least 70% accuracy on the validation set within 10 epochs. You can use the built-in train function, the training_part3 function from above, or implement your own training loop.

Describe what you did at the end of the notebook.

Some things you can try:
Filter size: Above we used 5x5 and 3x3; is this optimal?
Number of filters: Above we used 16 and 32 filters. Would more or fewer do better?
Pooling: We didn't use any pooling above. Would this improve the model?
Normalization: Would your model be improved with batch normalization, layer normalization, group normalization, or some other normalization strategy?
Network architecture: The ConvNet above has only three layers of trainable parameters. Would a deeper model do better?
Global average pooling: Instead of flattening after the final convolutional layer, would global average pooling do better? This strategy is used for example in Google's Inception network and in Residual Networks.
Regularization: Would some kind of regularization improve performance? Maybe weight decay or dropout?
NOTE: Batch Normalization / Dropout
If you are using Batch Normalization and Dropout, remember to pass is_training=True if you use the training_part3() function. BatchNorm and Dropout layers have different behaviors at training and inference time. training is a specific keyword argument reserved for this purpose in any tf.keras.Model's call() function. Read more about this here : https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/layers/BatchNormalization#methods https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/layers/Dropout#methods

Tips for training
For each network architecture that you try, you should tune the learning rate and other hyperparameters. When doing this there are a couple important things to keep in mind:

If the parameters are working well, you should see improvement within a few hundred iterations
Remember the coarse-to-fine approach for hyperparameter tuning: start by testing a large range of hyperparameters for just a few training iterations to find the combinations of parameters that are working at all.
Once you have found some sets of parameters that seem to work, search more finely around these parameters. You may need to train for more epochs.
You should use the validation set for hyperparameter search, and save your test set for evaluating your architecture on the best parameters as selected by the validation set.
Going above and beyond
If you are feeling adventurous there are many other features you can implement to try and improve your performance. You are not required to implement any of these, but don't miss the fun if you have time!

Alternative optimizers: you can try Adam, Adagrad, RMSprop, etc.
Alternative activation functions such as leaky ReLU, parametric ReLU, ELU, or MaxOut.
Model ensembles
Data augmentation
New Architectures
ResNets where the input from the previous layer is added to the output.
DenseNets where inputs into previous layers are concatenated together.
Have fun and happy training!

In [27]:
# Test 1 Architecture

def model_init_fn_1():
    model = None
    ############################################################################
    # TODO: Construct a three-layer ConvNet using tf.keras.Sequential.         #
    ############################################################################
    # *****START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    
    input_shape = (32, 32, 3)
    channel_1, channel_2, num_classes = 32, 16, 10
    initializer = tf.initializers.VarianceScaling(scale=2.0)
    layers = [
        tf.keras.layers.Conv2D(channel_1, (5, 5), 
                               padding='same', activation='relu', 
                               kernel_initializer=initializer,
                               input_shape=input_shape),
        tf.keras.layers.Conv2D(channel_2, (3, 3), 
                               padding='same', activation='relu', 
                               kernel_initializer=initializer),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(num_classes, activation='softmax',
                              kernel_initializer=initializer),
    ]
    model = tf.keras.Sequential(layers) 

    # *****END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    ############################################################################
    #                            END OF YOUR CODE                              #
    ############################################################################
    return model

In [28]:
# Test 2 Architecture

def model_init_fn_2():
    model = None
    ############################################################################
    # TODO: Construct a three-layer ConvNet using tf.keras.Sequential.         #
    ############################################################################
    # *****START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    
    input_shape = (32, 32, 3)
    channel_1, channel_2, num_classes = 32, 16, 10
    initializer = tf.initializers.VarianceScaling(scale=2.0)
    layers = [
        tf.keras.layers.Conv2D(channel_1, (5, 5), 
                               padding='same', activation='relu', 
                               kernel_initializer=initializer, 
                               input_shape=input_shape),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
        tf.keras.layers.Conv2D(channel_2, (3, 3), 
                               padding='same', activation='relu', 
                               kernel_initializer=initializer),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(512, activation='relu', 
                              kernel_initializer=initializer),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(num_classes, activation='softmax', 
                              kernel_initializer=initializer)
    ]
    model = tf.keras.Sequential(layers) 

    # *****END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    ############################################################################
    #                            END OF YOUR CODE                              #
    ############################################################################
    return model

In [29]:
# Test 3 Architecture

def model_init_fn_3():
    model = None
    ############################################################################
    # TODO: Construct a three-layer ConvNet using tf.keras.Sequential.         #
    ############################################################################
    # *****START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    
    input_shape = (32, 32, 3)
    channel_1, channel_2, num_classes = 32, 16, 10
    weight_decay = 1e-4
    dropout_rate = 0.5

    regularizer = tf.keras.regularizers.l2(weight_decay)
    layers = [
        tf.keras.layers.Conv2D(channel_1, (5, 5), padding='same', activation='relu',
                               kernel_initializer='he_normal',
                               kernel_regularizer=regularizer,
                               input_shape=input_shape),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
        tf.keras.layers.Conv2D(channel_2, (3, 3), padding='same', activation='relu',
                               kernel_initializer='he_normal',
                               kernel_regularizer=regularizer),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(512, activation='relu',
                             kernel_initializer='he_normal',
                             kernel_regularizer=regularizer),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dropout(dropout_rate),  # Add dropout
        tf.keras.layers.Dense(num_classes, activation='softmax',
                             kernel_initializer='he_normal',
                             kernel_regularizer=regularizer)
    ]
    model = tf.keras.Sequential(layers) 

    # *****END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    ############################################################################
    #                            END OF YOUR CODE                              #
    ############################################################################
    return model

In [30]:
# Test 4 Architecture

def model_init_fn_4():
    model = None
    ############################################################################
    # TODO: Construct a three-layer ConvNet using tf.keras.Sequential.         #
    ############################################################################
    # *****START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    
    input_shape = (32, 32, 3)
    channel_1, channel_2, num_classes = 32, 16, 10
    weight_decay = 1e-4
    dropout_rate = 0.5

    regularizer = tf.keras.regularizers.l2(weight_decay)
    layers = [
        tf.keras.layers.Conv2D(channel_1, (5, 5), padding='same', activation='relu',
                               kernel_initializer='he_normal',
                               kernel_regularizer=regularizer,
                               input_shape=input_shape),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
        tf.keras.layers.Conv2D(channel_2, (3, 3), padding='same', activation='relu',
                               kernel_initializer='he_normal',
                               kernel_regularizer=regularizer),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(512, activation='relu',
                             kernel_initializer='he_normal',
                             kernel_regularizer=regularizer),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dropout(dropout_rate),  # Add dropout
        tf.keras.layers.Dense(256, activation='relu',  # Add an additional hidden layer
                             kernel_initializer='he_normal',
                             kernel_regularizer=regularizer),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dropout(dropout_rate),  # Add dropout to the new hidden layer
        tf.keras.layers.Dense(num_classes, activation='softmax',
                             kernel_initializer='he_normal',
                             kernel_regularizer=regularizer)
    ]
    model = tf.keras.Sequential(layers) 

    # *****END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    ############################################################################
    #                            END OF YOUR CODE                              #
    ############################################################################
    return model

In [31]:
# Test 5 Architecture

def model_init_fn_5(dropout_rate=0.5):
    model = None
    ############################################################################
    # TODO: Construct a three-layer ConvNet using tf.keras.Sequential.         #
    ############################################################################
    # *****START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    
    input_shape = (32, 32, 3)
    channel_1, channel_2, num_classes = 32, 16, 10
    initializer = tf.initializers.VarianceScaling(scale=2.0)
    layers = [
        tf.keras.layers.Conv2D(channel_1, (5, 5), 
                               padding='same', activation='relu', 
                               kernel_initializer=initializer, 
                               input_shape=input_shape),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
        tf.keras.layers.Dropout(dropout_rate),  # Add dropout here
        tf.keras.layers.Conv2D(channel_2, (3, 3), 
                               padding='same', activation='relu', 
                               kernel_initializer=initializer),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
        tf.keras.layers.Dropout(dropout_rate),  # Add dropout here
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(512, activation='relu', 
                              kernel_initializer=initializer),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dropout(dropout_rate),  # Add dropout here
        tf.keras.layers.Dense(num_classes, activation='softmax', 
                              kernel_initializer=initializer)
    ]
    model = tf.keras.Sequential(layers) 

    # *****END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    ############################################################################
    #                            END OF YOUR CODE                              #
    ############################################################################
    return model

In [32]:
# Test 6 Architecture

def model_init_fn_6():
    model = None
    ############################################################################
    # TODO: Construct a three-layer ConvNet using tf.keras.Sequential.         #
    ############################################################################
    # *****START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    
    input_shape = (32, 32, 3)
    channel_1, channel_2, num_classes = 64, 32, 10
    weight_decay = 1e-4
    dropout_rate = 0.5

    regularizer = tf.keras.regularizers.l2(weight_decay)
    layers = [
        tf.keras.layers.Conv2D(channel_1, (5, 5), padding='same', activation='relu',
                               kernel_initializer='he_normal',
                               kernel_regularizer=regularizer,
                               input_shape=input_shape),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
        tf.keras.layers.Conv2D(channel_2, (3, 3), padding='same', activation='relu',
                               kernel_initializer='he_normal',
                               kernel_regularizer=regularizer),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(512, activation='relu',
                             kernel_initializer='he_normal',
                             kernel_regularizer=regularizer),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dropout(dropout_rate),  # Add dropout
        tf.keras.layers.Dense(num_classes, activation='softmax',
                             kernel_initializer='he_normal',
                             kernel_regularizer=regularizer)
    ]
    model = tf.keras.Sequential(layers) 

    # *****END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)*****
    ############################################################################
    #                            END OF YOUR CODE                              #
    ############################################################################
    return model

## Initial Tests

In [34]:
models = [
    model_init_fn_1(),
    model_init_fn_2(),
    model_init_fn_3(),
    model_init_fn_4(),
    model_init_fn_5(),
    model_init_fn_6()
]
history = []

for i, model in enumerate(models):
    print(f"Testing config {i}")
    print("-" * 20)
    learning_rate = 1e-3
    optimizer = tf.keras.optimizers.Adam(learning_rate=learning_rate)
    model.compile(optimizer=optimizer,
                  loss='sparse_categorical_crossentropy',
                  metrics=[tf.keras.metrics.sparse_categorical_accuracy])
    hist = model.fit(X_train, y_train, batch_size=64, epochs=5, validation_data=(X_val, y_val))
    model.evaluate(X_test, y_test)
    history.append(hist)
    print("*" * 20)

Testing config 0
--------------------
Epoch 1/5
766/766 [==============================] - 33s 41ms/step - loss: 1.5028 - sparse_categorical_accuracy: 0.4764 - val_loss: 1.2085 - val_sparse_categorical_accuracy: 0.5680
Epoch 2/5
766/766 [==============================] - 26s 34ms/step - loss: 1.0485 - sparse_categorical_accuracy: 0.6347 - val_loss: 1.1004 - val_sparse_categorical_accuracy: 0.6160
Epoch 3/5
766/766 [==============================] - 26s 34ms/step - loss: 0.8462 - sparse_categorical_accuracy: 0.7068 - val_loss: 1.1381 - val_sparse_categorical_accuracy: 0.6130
Epoch 4/5
766/766 [==============================] - 27s 35ms/step - loss: 0.7000 - sparse_categorical_accuracy: 0.7588 - val_loss: 1.1827 - val_sparse_categorical_accuracy: 0.6090
Epoch 5/5
313/313 [==============================] - 5s 15ms/step - loss: 1.3021 - sparse_categorical_accuracy: 0.6051
********************
Testing config 1
--------------------
Epoch 1/5
766/766 [==============================] - 34s 43m

Initial Test Notes (5 epochs): 
- Test 1: (~60%)
    - Adam optimizer with 1e-3 learning rate
- Test 2: (~70%)
    - Adam optimizer with 1e-3 learning rate
    - Adding BatchNormalization and MaxPooling 
    - Increased accruacy to - with extra layer
    - Explicitly define optimizer with learning rate of 1e-3 
- Test 3: (~70%)
    - Adam optimizer with 1e-3 learning rate
    - Add regularization
    - Add dropout
    - Add another layer
- Test 4: (~70%)
    - Adam optimizer with 1e-3 learning rate
    - Add another layer with dropout and regularization
- Test 5: (64%)
    - Adam optimizer with 1e-3 learning rate
    - Go back to test 2 architecture and add just dropout and another layer
- Test 6: (~72%)
    - Add regularization
    - Add BatchNormalization
    - Add MaxPooling2D
    - Add dropout
    - Add another layer 
    - change channel_1, channel_2 = 64, 32

## Observations

So far the best configurations were from architecture 2, 3, 4, and 5. The next series of tests will be for 10 epochs to see which configuration will perform the best.

## Final Tests

In [35]:
models = [
    model_init_fn_2(),
    model_init_fn_3(),
    model_init_fn_4(),
    model_init_fn_6()
]
history = []

for i, model in enumerate(models):
    print(f"Testing config {i}")
    print("-" * 20)
    learning_rate = 1e-3
    optimizer = tf.keras.optimizers.Adam(learning_rate=learning_rate)
    model.compile(optimizer=optimizer,
                  loss='sparse_categorical_crossentropy',
                  metrics=[tf.keras.metrics.sparse_categorical_accuracy])
    hist = model.fit(X_train, y_train, batch_size=64, epochs=10, validation_data=(X_val, y_val))
    model.evaluate(X_test, y_test)
    history.append(hist)
    print("*" * 20)

Testing config 0
--------------------
Epoch 1/10
766/766 [==============================] - 20s 25ms/step - loss: 1.3170 - sparse_categorical_accuracy: 0.5456 - val_loss: 1.0202 - val_sparse_categorical_accuracy: 0.6360
Epoch 2/10
766/766 [==============================] - 29s 38ms/step - loss: 0.9205 - sparse_categorical_accuracy: 0.6758 - val_loss: 0.9570 - val_sparse_categorical_accuracy: 0.6550
Epoch 3/10
766/766 [==============================] - 29s 38ms/step - loss: 0.7492 - sparse_categorical_accuracy: 0.7394 - val_loss: 0.9295 - val_sparse_categorical_accuracy: 0.7020
Epoch 4/10
766/766 [==============================] - 29s 38ms/step - loss: 0.6105 - sparse_categorical_accuracy: 0.7862 - val_loss: 0.9334 - val_sparse_categorical_accuracy: 0.6890
Epoch 5/10
766/766 [==============================] - 41s 53ms/step - loss: 0.4851 - sparse_categorical_accuracy: 0.8327 - val_loss: 0.9612 - val_sparse_categorical_accuracy: 0.7050
Epoch 6/10
766/766 [==============================] 

Initial Test Notes (10 epochs): 
- Test 2: (~68%)
    - Adam optimizer with 1e-3 learning rate
    - Adding BatchNormalization and MaxPooling 
    - Increased accruacy to - with extra layer
    - Explicitly define optimizer with learning rate of 1e-3 
- Test 3: (~73%)
    - Adam optimizer with 1e-3 learning rate
    - Add regularization
    - Add dropout
    - Add another layer
- Test 4: (~73%)
    - Adam optimizer with 1e-3 learning rate
    - Add another layer with dropout and regularization
- Test 6: (~74%)
    - Adam optimizer with 1e-3 learning rate
    - Add regularization
    - Add BatchNormalization
    - Add MaxPooling2D
    - Add dropout
    - Add another layer 
    - change channel_1, channel_2 = 64, 32

## Observations

70% accuracy on the validation set within 10 epochs was achieved on 3 of the 4 architectures listed above. The best performing architecture was from test 6 with a validation accruacy of roughly 74%. The modification can be seen in the list under Test 6 above.